### ISSUE TITLE
propagate clusterRoleBindings objects

### ISSUE THREAD

**Issue by zfrhv** [kubernetes-sigs/hierarchical-namespaces#328]
We are using openshift, and our client has anyuid securityContextConstraint on the service account in the namespace (which is simply clusterRoleBinding for the service account)

When he creates sub-namespaces, the clusterRoleBinding of anyuid is not being propagated, thus his application can't run.

Is there any way to make it possible propagate cluster-wide objects?
maybe add field in HNCConfiguration that specifies which field to change when propagating cluster-wide object, or maybe there is other workaround?

**Comment by zfrhv** [comment]
something like this:
```yaml
apiVersion: hnc.x-k8s.io/v1alpha2
kind: HNCConfiguration
metadata:
  name: config
spec:
  resources:
  - group: rbac.authorization.k8s.io
    resource: clusterrolebindings
    namespaceField: subjects[0].namespace
```
and the name of the propogated object can be randomly generated + label of the namespace.

for example clusterRoleBinding with name test can become test-wgkrvk with label `hnc.x-k8s.io/namespace: child1`
```

**Comment by zfrhv** [comment]
maybe even `namespaceFields` can be an array, and then you can specify multiple paths.

its also good even for namespaced custom resources which have other places where the namespace is mentioned.

let me know what do you think and i will try to make a PR

**Comment by erikgb** [comment]
I don't think this is a problem HNC should solve. SCC is a cluster-scoped resource type. If you need the `anyuid` SCC added to all serviceaccounts, I think you should configure your cluster to do it. But I don't recommend it, as it will break the OpenShift "secure by default" policy.

**Comment by zfrhv** [comment]
hii @erikgb thanks for the reply
i need to give scc only to a specific client and his sub-namespaces

also i was thinking that there may be other objects that somehow related to a namespace and should be propogated

for example OperatorGroup, would be nice to propogate as well. like this:
```yaml
apiVersion: hnc.x-k8s.io/v1alpha2
kind: HNCConfiguration
metadata:
  name: config
spec:
  resources:
  - group: operators.coreos.com
    resource: operatorgroups
    namespaceField: spec.targetNamespaces[0]
```
operatorGroup is a namespaced object, but its not being propogated properly.

so i thought that we can configure the propogation field, and this will also allow us to know how a cluster-wide object can be propogated.
which will solve the clusterRoleBinding problem

there may be other CRDs (cluster-wide or not) that are related to namespace and should be propogated as well (that im not aware of).

**Comment by erikgb** [comment]
I am pretty sure HNC should be able to propagate any namespaced resource - if configured to do so. If this is not the behavior, please report an issue.

But Openshift SCC is a cluster-scoped resource, so IMO **propagate** does not make sense - since it is cluster-scoped.

**Comment by zfrhv** [comment]
> If this is not the behavior, please report an issue.

sounds fair, i will open another case

> But Openshift SCC is a cluster-scoped resource, so IMO propagate does not make sense - since it is cluster-scoped.

im not talking about scc, im talking about cluster wide permission (clusterRoleBinding), if a service account has permissions on some cluster-wide object (like list storageClasses, namespaces, nodes...) then sounds fair to have similar permissions in the sub-namespaces.

i agree that cluster-wide object should be created only once and not per each namespace, but what can be done if cluster-wide permissions cant be namespaced, how to propogate those permissions?

but you have a good point, maybe adding multiple `subjects` in the clusterRoleBinding is better than creating multiple clusterRoleBindings for each namespace.

**Comment by zfrhv** [comment]
#329

**Comment by erikgb** [comment]
> but you have a good point, maybe adding multiple `subjects` in the clusterRoleBinding is better than creating multiple clusterRoleBindings for each namespace.

Yes, that is one way to solve it. Granting access to elevated SCCs is a cluster-admin responsibility IMO.

**Comment by zfrhv** [comment]
Sorry I missed the notification somehow

> Granting access to elevated SCCs is a cluster-admin responsibility IMO.

Yes but there is no way to automate it.
If I could configure the HNCConfiguration to propagate the clusterRoleBinding I would consider it as I'm responsible for it (as long as its propagated as expected)

Seems to me that you dont like the way SCC works, and I dont like it too, but this is an issue about propagation of cluster-wide permissions.

For example (real scenario) I allow my client to list nodes with their application by giving CluterRoleBinding to their serviceAccount:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: node-list-client1
subjects:
- kind: ServiceAccount
  name: client1-sa
  namespace: client1
roleRef:
  kind: ClusterRole
  name: node-list-role
  apiGroup: rbac.authorization.k8s.io
```
The client has pipeline that automatically creates child subNamespace, creates the application in that child subNamespace, runs tests, and deletes the subNamespace.
Seems like all was good because all the permissions were namespace (roleBindings propagated well).
But then we realized that the application requires to list nodes, and now we are stuck because HNC cant propagate ClusterRoleBinding, and I cant give the client permissions to create ClusterRoleBindings.

**Comment by erikgb** [comment]
I think I understand what you wish for, but ClusterRoleBinding is a cluster-scoped resource. On what basis should HNC propagate something that is not namespaced?

And you are correct, I am not a big fan of SCC. 😉

**Comment by zfrhv** [comment]
> On what basis should HNC propagate something that is not namespaced?

I don't really have a good reason why HNC should propagate cluster-wide objects but that was my logic:
- My only solution is to write a script that detects the ClusterRoleBinding and knows how to propagate it
- Why to create a whole new application if this script can be part of HNC, and many users will probably face the same issue and will need this feature
- Why to add propagation only for ClusterRoleBinding and not generally cluster-wide objects, sounds like easy to implement and opens many possibilities (unlimited, the same way as kubernetes works with anything however you like)

**Comment by adrianludwin** [comment]
Sorry I missed this (thanks @erikgb for digging in) - I think the issue is that you'd like _service accounts_ in a child namespace to inherit the permissions from the SA in the parent namespace - right?

A regular, namespaced `RoleBinding` can refer to a `ClusterRole`, so maybe a solution is to put the `RoleBinding` in the root namespaces and let that be propagated? Then, as long as the service accounts have the same names, it should just work, right?

**Comment by zfrhv** [comment]
> I think the issue is that you'd like service accounts in a child namespace to inherit the permissions from the SA in the parent namespace - right?

yes

> A regular, namespaced RoleBinding can refer to a ClusterRole, so maybe a solution is to put the RoleBinding in the root namespaces and let that be propagated? Then, as long as the service accounts have the same names, it should just work, right?

I'm pretty confident it wont work because when using roleBinding it can give permissions only in the namespace, even when using a clusterRole.  
For example if I gave cluster-admin to a serviceAccount in root namespace, its just not possible to propagate it with roleBinding to child namespace, clusterRoleBinding must be used.

But I will still try to use the roleBinding for listing nodes and other permissions 👍

**Comment by zfrhv** [comment]
hello,
yes indeed the cluster-wide permissions can be used only with ClusterRoleBindings.  
I tried to give permission on nodes with roleBinding and it didnt worked.

so what do you think the solution could be?  
if you think that adding `namespaceFields` field to the `HNCConfiguration` CRD then let me know, I will study go and try to create a PR.

**Comment by adrianludwin** [comment]
Oh I see, I forgot that the permission you wanted to inherit was for
non-namespaced objects.

Rather than a general-purpose "copy cluster-wide objects," since we don't
actually know what rules would work well in a general-purpose way, I'd
prefer adding a ClusterRoleBinding-specific controller - it'll be less
work, easier to replace in the future, and less dangerous due to the
tighter scope. Do you want to write up a quick design doc for this?

**Comment by zfrhv** [comment]
> easier to replace in the future

You got me on that sentence 

> Do you want to write up a quick design doc for this?

No clue what to do, if I do it then it wont be quick.
I volunteer to create PR only because its my suggestion and other people seem busy.

But I can try, I wanna help more!

**Comment by adrianludwin** [comment]
> > easier to replace in the future
> 
> You got me on that sentence

If we create a feature called "cluster object propagation," and then it doesn't work well for future use cases, we need to create a "cluster object propagation v2" or try to slowly fix the first feature in a backwards-compatible way, which is hard. By contrast, if we create a clusterrolebinding-specific feature, and later discover more general-purpose use cases that makes us confident it will work for other objects, we can introduce cluster object propagation in the future, without necessarily affecting the clusterrolebinding, which can slowly be deprecated.

> > Do you want to write up a quick design doc for this?
> 
> No clue what to do, if I do it then it wont be quick. I volunteer to create PR only because its my suggestion and other people seem busy.
> 
> But I can try, I wanna help more!

In the design doc, cover the proposed API and behaviour, as well as any tricky implementation details. For example, one key decision would be "do I modify existing clusterrolebinding objects to add the new service accounts" or "do I create one propagated clusterrolebinding for every child namespace" or "one clusterrolebinding for _all_ descendant namespaces." You'll probably also want to cover how to disable the feature if necessary.

There's no specific formula for a design doc, just make a Google Doc with all your ideas so others can comment on it.

**Comment by zfrhv** [comment]
hello~

sorry it took a while, war started so we stopped doing anything we had

I made some doc, not too deep but im not sure whats required:
https://docs.google.com/document/d/1r6mr-OlfYt3yaLpHwX-oLlve9Fs438tddoB7a-VL7yg/edit?usp=drivesdk

no new apis are required, just changing already existing apis.
Is that wrong? should I try to seperate it so it would be easier to replace in the future?

i could make it more generic in case there would be other clusterWide objects in the future, but you already convinced me that its a workaround so stricting it to only clusterRoleBinding sounds right.

I dont know GOLang nor how this operator works, so I think doodling some PR will work better rather than just guessing what to modify just by looking at the code.

**Comment by adrianludwin** [comment]
Hey, very sorry to hear. The world's so broken I can't even be sure which
war you're talking about, but I hope that you and yours are out of harm's
way. I'll review the doc when I get a chance, likely this Friday. Stay safe.

**Comment by zfrhv** [comment]
hello~

sorry for the undetailed google doc, im not sure myself how to implement clusterRole propagation

i think it will get to know better why attempting to create a PR

can i try to create a PR?

**Comment by zfrhv** [comment]
hello~
im very sorry but it turns out that we can use role binding for SCC instead of clusterRoleBinding

so this feature is not needed for me anymore, thus i will close it

very sorry
